link/netkit_test.go

```diff
@@ -0,0 +1,89 @@
+package link
+
+import (
+	"fmt"
+	"net"
+	"testing"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal/testutils"
+
+	"github.com/go-quicktest/qt"
+	"github.com/vishvananda/netlink"
+)
+
+func TestAttachNetkit(t *testing.T) {
+	testutils.SkipOnOldKernel(t, "6.7", "Netkit Device")
+
+	prog := mustLoadProgram(t, ebpf.SchedCLS, ebpf.AttachNetkitPrimary, "")
+	link, _ := mustAttachNetkit(t, prog, ebpf.AttachNetkitPrimary)
+
+	testLink(t, link, prog)
+}
+
+func TestNetkitAnchor(t *testing.T) {
+	testutils.SkipOnOldKernel(t, "6.7", "Netkit Device")
+
+	a := mustLoadProgram(t, ebpf.SchedCLS, ebpf.AttachNetkitPrimary, "")
+	b := mustLoadProgram(t, ebpf.SchedCLS, ebpf.AttachNetkitPrimary, "")
+
+	linkA, iface := mustAttachNetkit(t, a, ebpf.AttachNetkitPrimary)
+
+	programInfo, err := a.Info()
+	qt.Assert(t, qt.IsNil(err))
+	programID, _ := programInfo.ID()
+
+	linkInfo, err := linkA.Info()
+	qt.Assert(t, qt.IsNil(err))
+	linkID := linkInfo.ID
+
+	for _, anchor := range []Anchor{
+		Head(),
+		Tail(),
+		BeforeProgram(a),
+		BeforeProgramByID(programID),
+		AfterLink(linkA),
+		AfterLinkByID(linkID),
+	} {
+		t.Run(fmt.Sprintf("%T", anchor), func(t *testing.T) {
+			linkB, err := AttachNetkit(NetkitOptions{
+				Program:   b,
+				Attach:    ebpf.AttachNetkitPrimary,
+				Interface: iface,
+				Anchor:    anchor,
+			})
+			qt.Assert(t, qt.IsNil(err))
+			qt.Assert(t, qt.IsNil(linkB.Close()))
+		})
+	}
+}
+
+func mustAttachNetkit(tb testing.TB, prog *ebpf.Program, attachType ebpf.AttachType) (Link, int) {
+	var err error
+	nk := &netlink.Netkit{
+		LinkAttrs: netlink.LinkAttrs{
+			Name: "nk1",
+		},
+		Mode:       netlink.NETKIT_MODE_L2,
+		Policy:     netlink.NETKIT_POLICY_FORWARD,
+		PeerPolicy: netlink.NETKIT_POLICY_BLACKHOLE,
+	}
+	err = netlink.LinkAdd(nk)
```

Comment:
How much work would it be to use the library Timo mentioned, or something else deriving from https://pkg.go.dev/github.com/mdlayher/netlink? visvhananda/netlink is quite low quality.

Reply:
I don't think `mdlayher/netlink` has support for netkit yet though ? Will need to add support for it first IIUC.

Reply:
If we strongly feel about this, I can take a look at adding support for this.

Reply:
> If we strongly feel about this

We do. :) The goal is to start moving parts of Cilium off of v/netlink once we've done some refactoring because it keeps causing problems.

Reply:
Marking this resolved since its now tracked with https://github.com/cilium/ebpf/issues/1350